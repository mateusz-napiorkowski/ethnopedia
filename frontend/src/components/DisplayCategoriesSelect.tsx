import React, { useState, useEffect, useRef } from "react";

export type Option = {
    value: string;
    label: string;
};

type DisplayCategoriesSelectProps = {
    selectedDisplayCategories: string[];
    setSelectedDisplayCategories: (categories: string[]) => void;
    categoryOptions: Option[];
    customOptions: Option[];
    formatOptionLabel: (option: Option, context: { context: string }) => React.ReactNode;
};

const DisplayCategoriesSelect: React.FC<DisplayCategoriesSelectProps> = ({
                                                                             selectedDisplayCategories,
                                                                             setSelectedDisplayCategories,
                                                                             categoryOptions,
                                                                             customOptions,
                                                                             formatOptionLabel,
                                                                         }) => {
    const [open, setOpen] = useState<boolean>(false);
    const dropdownRef = useRef<HTMLDivElement>(null);

    // Zamykamy dropdown przy kliknięciu poza obszarem komponentu
    useEffect(() => {
        const handleClickOutside = (event: MouseEvent) => {
            if (dropdownRef.current && !dropdownRef.current.contains(event.target as Node)) {
                setOpen(false);
            }
        };
        document.addEventListener("mousedown", handleClickOutside);
        return () => document.removeEventListener("mousedown", handleClickOutside);
    }, []);

    // Funkcja przełączająca zaznaczenie pojedynczej kategorii
    const toggleCategory = (categoryValue: string) => {
        if (selectedDisplayCategories.includes(categoryValue)) {
            // Zapobiegamy usunięciu ostatniej zaznaczonej kategorii
            if (selectedDisplayCategories.length === 1) return;
            setSelectedDisplayCategories(selectedDisplayCategories.filter((cat) => cat !== categoryValue));
        } else {
            setSelectedDisplayCategories([...selectedDisplayCategories, categoryValue]);
        }
    };

    // Obsługa opcji specjalnych
    const handleSpecialOption = (option: Option) => {
        if (option.value === "select_all") {
            const allCategories = categoryOptions.map((opt) => opt.value);
            setSelectedDisplayCategories(allCategories);
        } else if (option.value === "deselect_all") {
            const defaultSelection = categoryOptions.length ? [categoryOptions[0].value] : [];
            setSelectedDisplayCategories(defaultSelection);
        }
    };

    // Renderowanie pojedynczej opcji
    const renderOptionItem = (option: Option) => {
        if (option.value === "select_all" || option.value === "deselect_all") {
            return (
                <div
                    key={option.value}
                    onClick={() => {
                        handleSpecialOption(option);
                        setOpen(false);
                    }}
                    className="cursor-pointer py-2 px-4 border-b border-gray-300"
                >
                    {formatOptionLabel(option, { context: "menu" })}
                </div>
            );
        } else {
            const isChecked = selectedDisplayCategories.includes(option.value);
            return (
                <div
                    key={option.value}
                    onClick={() => toggleCategory(option.value)}
                    className="cursor-pointer flex items-center py-2 px-4 border-b border-gray-300"
                >
                    <input
                        type="checkbox"
                        checked={isChecked}
                        readOnly
                        className="mr-2"
                    />
                    <span>{option.label}</span>
                </div>
            );
        }
    };

    return (
        <div className="relative" ref={dropdownRef}>
            {/* Pole wyboru – wyświetla liczbę zaznaczonych kategorii*/}
            <div
                onClick={() => setOpen(!open)}
                className="cursor-pointer py-2 px-4 border bg-white dark:bg-gray-800 border-gray-300 rounded-lg text-sm flex items-center justify-between w-full"
            >
    <span className="pr-4">
        {selectedDisplayCategories.length > 0
            ? `${selectedDisplayCategories.length}`
            : "Wybierz kategorię"}
    </span>
                {/* Strzałka */}
                <svg
                    className={`h-4 w-4 transition-transform duration-200 ${open ? "transform rotate-180" : ""}`}
                    fill="none"
                    stroke="currentColor"
                    viewBox="0 0 24 24"
                >
                    <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M19 9l-7 7-7-7"/>
                </svg>
            </div>
            {/* Rozwijana lista */}
            {open && (
                <div
                    className="absolute z-10 bg-white dark:bg-gray-800 border border-gray-300 mt-1 rounded shadow-md max-h-60 overflow-y-auto"
                    style={{minWidth: "100%", width: "auto"}}
                >
                    {customOptions.map((option) => renderOptionItem(option))}
                </div>
            )}
        </div>
    );
};

export default DisplayCategoriesSelect;
