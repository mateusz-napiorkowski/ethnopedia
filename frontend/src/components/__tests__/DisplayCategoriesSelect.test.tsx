import '@testing-library/jest-dom';
import { render } from '@testing-library/react'
import DisplayCategoriesSelect from "../DisplayCategoriesSelect"
import { MemoryRouter, Routes, Route } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import { useState } from 'react';

const user = userEvent.setup()

const exampleCollectionId = "67f84d80d2ac8e9a1e67cca4"
const renderComponent = (collectionId = exampleCollectionId) => {
    const Wrapper = () => {
        const [selectedDisplayCategories, setSelectedDisplayCategories] = useState(["Tytuł", "Tytuł.Podtytuł", "Artyści"])
        
        const categoryOptions = [
            {
                "value": "Tytuł",
                "label": "Tytuł"
            },
            {
                "value": "Tytuł.Podtytuł",
                "label": "Tytuł.Podtytuł"
            },
            {
                "value": "Artyści",
                "label": "Artyści"
            },
            {
                "value": "Rok",
                "label": "Rok"
            }
        ]

        const customOptions = [
            {
                "value": "select_all",
                "label": "Zaznacz wszystkie"
            },
            {
                "value": "deselect_all",
                "label": "Odznacz wszystkie"
            },
            {
                "value": "Tytuł",
                "label": "Tytuł"
            },
            {
                "value": "Tytuł.Podtytuł",
                "label": "Tytuł.Podtytuł"
            },
            {
                "value": "Artyści",
                "label": "Artyści"
            },
            {
                "value": "Rok",
                "label": "Rok"
            }
        ]

        type Option = {
            value: string;
            label: string;
        };

        const formatOptionLabel = (option: Option, { context }: { context: string }) => {
            if (context === "menu") {
                if (option.value === "select_all" || option.value === "deselect_all") {
                    return (
                        <div
                            className="text-gray-500 dark:text-gray-300 underline"
                        >
                            {option.label}
                        </div>
                    );
                }
            }
            return option.label;
        }

        return (<DisplayCategoriesSelect
            selectedDisplayCategories={selectedDisplayCategories}
            setSelectedDisplayCategories={setSelectedDisplayCategories}
            categoryOptions={categoryOptions}
            customOptions={customOptions}
            formatOptionLabel={formatOptionLabel}
        />

        )
    }
    
    return render(
        <MemoryRouter initialEntries={[`/collections/${collectionId}/artworks/`]}>
            <Routes>
                <Route path="/collections/:collectionId/artworks/" element={<Wrapper/>}/>
            </Routes>  
        </MemoryRouter>
    );
};

describe("DisplayCategoriesSelect tests", () => {
    beforeEach(() => {
        jest.clearAllMocks()
    });

    it("should render initial state", async () => {           
        const { container } = renderComponent()
        
        expect(container).toMatchSnapshot()
    })

    it("should expand select menu when component is clicked", async () => {           
        const {container, getByLabelText} = renderComponent()

        await user.click(getByLabelText("open/close-display-categories-select"))

        expect(container).toMatchSnapshot()
    })

    it("should hide select menu when user clicks outside of component", async () => {           
        const {container, getByLabelText} = renderComponent()

        await user.click(getByLabelText("open/close-display-categories-select"))
        await user.click(document.body)

        expect(container).toMatchSnapshot()
    })

    it("should uncheck all categories except first when deselect all button is clicked", async () => {           
        const {container, getByLabelText, getByText} = renderComponent()
        await user.click(getByLabelText("open/close-display-categories-select"))
        const deselectAllButton = getByText(/odznacz wszystkie/i)

        await user.click(deselectAllButton)
        await user.click(getByLabelText("open/close-display-categories-select"))

        expect(container).toMatchSnapshot()
    })

    it("should check all categories when select all button is clicked", async () => {           
        const {container, getByLabelText, getByText} = renderComponent()
        await user.click(getByLabelText("open/close-display-categories-select"))
        const selectAllButton = getByText(/zaznacz wszystkie/i)

        await user.click(selectAllButton)
        await user.click(getByLabelText("open/close-display-categories-select"))

        expect(container).toMatchSnapshot()
    })

    it("should check/uncheck categories when appropriate options are clicked", async () => {           
        const {getByLabelText} = renderComponent()
        await user.click(getByLabelText("open/close-display-categories-select"))
        const titleOption = getByLabelText("Tytuł-check/uncheck")
        const subtitleOption = getByLabelText("Tytuł.Podtytuł-check/uncheck")
        const artistsOption = getByLabelText("Artyści-check/uncheck")
        const yearOption = getByLabelText("Rok-check/uncheck")

        await user.click(yearOption)
        await user.click(titleOption)
        await user.click(titleOption)
        await user.click(subtitleOption)
        await user.click(yearOption)
        await user.click(subtitleOption)
        await user.click(yearOption)
        await user.click(yearOption)
        await user.click(artistsOption)

        expect(titleOption).toBeChecked()
        expect(subtitleOption).toBeChecked()
        expect(artistsOption).not.toBeChecked()
        expect(yearOption).not.toBeChecked()
    })
})