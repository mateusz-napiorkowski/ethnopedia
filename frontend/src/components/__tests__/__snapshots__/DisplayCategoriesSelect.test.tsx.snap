// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DisplayCategoriesSelect tests should check all categories when select all button is clicked 1`] = `
<div>
  <div
    class="relative"
  >
    <div
      aria-label="open/close-display-categories-select"
      class="cursor-pointer py-2 px-4 border bg-white dark:bg-gray-800 border-gray-300 rounded-lg text-sm flex items-center justify-between w-full"
    >
      <span
        class="pr-4"
      >
        4
      </span>
      <svg
        class="h-4 w-4 transition-transform duration-200 transform rotate-180"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 9l-7 7-7-7"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
    <div
      class="absolute z-10 bg-white dark:bg-gray-800 border border-gray-300 mt-1 rounded shadow-md max-h-60 overflow-y-auto"
      data-testid="DisplayCategoriesSelectExpanded"
      style="min-width: 100%; width: auto;"
    >
      <div
        class="cursor-pointer py-2 px-4 border-b border-gray-300"
      >
        <div
          class="text-gray-500 dark:text-gray-300 underline"
        >
          Zaznacz wszystkie
        </div>
      </div>
      <div
        class="cursor-pointer py-2 px-4 border-b border-gray-300"
      >
        <div
          class="text-gray-500 dark:text-gray-300 underline"
        >
          Odznacz wszystkie
        </div>
      </div>
      <div
        class="cursor-pointer flex items-center py-2 px-4 border-b border-gray-300"
      >
        <input
          aria-label="Tytuł-check/uncheck"
          checked=""
          class="mr-2"
          readonly=""
          type="checkbox"
        />
        <span>
          Tytuł
        </span>
      </div>
      <div
        class="cursor-pointer flex items-center py-2 px-4 border-b border-gray-300"
      >
        <input
          aria-label="Tytuł.Podtytuł-check/uncheck"
          checked=""
          class="mr-2"
          readonly=""
          type="checkbox"
        />
        <span>
          Tytuł.Podtytuł
        </span>
      </div>
      <div
        class="cursor-pointer flex items-center py-2 px-4 border-b border-gray-300"
      >
        <input
          aria-label="Artyści-check/uncheck"
          checked=""
          class="mr-2"
          readonly=""
          type="checkbox"
        />
        <span>
          Artyści
        </span>
      </div>
      <div
        class="cursor-pointer flex items-center py-2 px-4 border-b border-gray-300"
      >
        <input
          aria-label="Rok-check/uncheck"
          checked=""
          class="mr-2"
          readonly=""
          type="checkbox"
        />
        <span>
          Rok
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`DisplayCategoriesSelect tests should expand select menu when component is clicked 1`] = `
<div>
  <div
    class="relative"
  >
    <div
      aria-label="open/close-display-categories-select"
      class="cursor-pointer py-2 px-4 border bg-white dark:bg-gray-800 border-gray-300 rounded-lg text-sm flex items-center justify-between w-full"
    >
      <span
        class="pr-4"
      >
        3
      </span>
      <svg
        class="h-4 w-4 transition-transform duration-200 transform rotate-180"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 9l-7 7-7-7"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
    <div
      class="absolute z-10 bg-white dark:bg-gray-800 border border-gray-300 mt-1 rounded shadow-md max-h-60 overflow-y-auto"
      data-testid="DisplayCategoriesSelectExpanded"
      style="min-width: 100%; width: auto;"
    >
      <div
        class="cursor-pointer py-2 px-4 border-b border-gray-300"
      >
        <div
          class="text-gray-500 dark:text-gray-300 underline"
        >
          Zaznacz wszystkie
        </div>
      </div>
      <div
        class="cursor-pointer py-2 px-4 border-b border-gray-300"
      >
        <div
          class="text-gray-500 dark:text-gray-300 underline"
        >
          Odznacz wszystkie
        </div>
      </div>
      <div
        class="cursor-pointer flex items-center py-2 px-4 border-b border-gray-300"
      >
        <input
          aria-label="Tytuł-check/uncheck"
          checked=""
          class="mr-2"
          readonly=""
          type="checkbox"
        />
        <span>
          Tytuł
        </span>
      </div>
      <div
        class="cursor-pointer flex items-center py-2 px-4 border-b border-gray-300"
      >
        <input
          aria-label="Tytuł.Podtytuł-check/uncheck"
          checked=""
          class="mr-2"
          readonly=""
          type="checkbox"
        />
        <span>
          Tytuł.Podtytuł
        </span>
      </div>
      <div
        class="cursor-pointer flex items-center py-2 px-4 border-b border-gray-300"
      >
        <input
          aria-label="Artyści-check/uncheck"
          checked=""
          class="mr-2"
          readonly=""
          type="checkbox"
        />
        <span>
          Artyści
        </span>
      </div>
      <div
        class="cursor-pointer flex items-center py-2 px-4 border-b border-gray-300"
      >
        <input
          aria-label="Rok-check/uncheck"
          class="mr-2"
          readonly=""
          type="checkbox"
        />
        <span>
          Rok
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`DisplayCategoriesSelect tests should hide select menu when user clicks outside of component 1`] = `
<div>
  <div
    class="relative"
  >
    <div
      aria-label="open/close-display-categories-select"
      class="cursor-pointer py-2 px-4 border bg-white dark:bg-gray-800 border-gray-300 rounded-lg text-sm flex items-center justify-between w-full"
    >
      <span
        class="pr-4"
      >
        3
      </span>
      <svg
        class="h-4 w-4 transition-transform duration-200 "
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 9l-7 7-7-7"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`DisplayCategoriesSelect tests should render initial state 1`] = `
<div>
  <div
    class="relative"
  >
    <div
      aria-label="open/close-display-categories-select"
      class="cursor-pointer py-2 px-4 border bg-white dark:bg-gray-800 border-gray-300 rounded-lg text-sm flex items-center justify-between w-full"
    >
      <span
        class="pr-4"
      >
        3
      </span>
      <svg
        class="h-4 w-4 transition-transform duration-200 "
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 9l-7 7-7-7"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`DisplayCategoriesSelect tests should uncheck all categories except first when deselect all button is clicked 1`] = `
<div>
  <div
    class="relative"
  >
    <div
      aria-label="open/close-display-categories-select"
      class="cursor-pointer py-2 px-4 border bg-white dark:bg-gray-800 border-gray-300 rounded-lg text-sm flex items-center justify-between w-full"
    >
      <span
        class="pr-4"
      >
        1
      </span>
      <svg
        class="h-4 w-4 transition-transform duration-200 transform rotate-180"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 9l-7 7-7-7"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
    <div
      class="absolute z-10 bg-white dark:bg-gray-800 border border-gray-300 mt-1 rounded shadow-md max-h-60 overflow-y-auto"
      data-testid="DisplayCategoriesSelectExpanded"
      style="min-width: 100%; width: auto;"
    >
      <div
        class="cursor-pointer py-2 px-4 border-b border-gray-300"
      >
        <div
          class="text-gray-500 dark:text-gray-300 underline"
        >
          Zaznacz wszystkie
        </div>
      </div>
      <div
        class="cursor-pointer py-2 px-4 border-b border-gray-300"
      >
        <div
          class="text-gray-500 dark:text-gray-300 underline"
        >
          Odznacz wszystkie
        </div>
      </div>
      <div
        class="cursor-pointer flex items-center py-2 px-4 border-b border-gray-300"
      >
        <input
          aria-label="Tytuł-check/uncheck"
          checked=""
          class="mr-2"
          readonly=""
          type="checkbox"
        />
        <span>
          Tytuł
        </span>
      </div>
      <div
        class="cursor-pointer flex items-center py-2 px-4 border-b border-gray-300"
      >
        <input
          aria-label="Tytuł.Podtytuł-check/uncheck"
          class="mr-2"
          readonly=""
          type="checkbox"
        />
        <span>
          Tytuł.Podtytuł
        </span>
      </div>
      <div
        class="cursor-pointer flex items-center py-2 px-4 border-b border-gray-300"
      >
        <input
          aria-label="Artyści-check/uncheck"
          class="mr-2"
          readonly=""
          type="checkbox"
        />
        <span>
          Artyści
        </span>
      </div>
      <div
        class="cursor-pointer flex items-center py-2 px-4 border-b border-gray-300"
      >
        <input
          aria-label="Rok-check/uncheck"
          class="mr-2"
          readonly=""
          type="checkbox"
        />
        <span>
          Rok
        </span>
      </div>
    </div>
  </div>
</div>
`;
